src/decoder/lattice-incremental-decoder.h

```diff
@@ -0,0 +1,448 @@
+  static void TopSortTokens(Token *tok_list, std::vector<Token *> *topsorted_list);
+
+  void ClearActiveTokens();
+
+  /// Obtains a CompactLattice for the part of this utterance that has been
+  /// decoded so far.  If you call this multiple times (calling it on every frame
+  /// would not make
+  /// sense, but every, say, 10, to 40 frames might make sense) it will spread out
+  /// the
+  /// work of determinization over time,which might be useful for online
+  /// applications.
+  ///
+  ///   @param [in]  use_final_probs  If true *and* at least one final-state in HCLG
+  ///                         was active on the final frame, include final-probs from
+  ///                         HCLG
+  ///                         in the lattice.  Otherwise treat all final-costs of
+  ///                         states active
+  ///                         on the most recent frame as zero (i.e.  Weight::One()).
+  ///   @param [in]  redeterminize    If true, re-determinize the CompactLattice
+  ///                         after appending the most recently decoded chunk to it,
+  ///                         to
+  ///                         ensure that the output is fully deterministic.
+  ///                         This does extra work, but not nearly as much as
+  ///                         determinizing
+  ///                          a RawLattice from scratch.
+  ///  @param [out] lat   The CompactLattice representing what has been decoded
+  ///                          so far.
+  ///   @return reached_final    This function will returns true if a state that was
+  ///   final in
+  ///                         HCLG was active on the most recent frame, and false
+  ///                         otherwise.
+  ///                         CAUTION: this is not the same meaning as the return
+  ///                         value of
+  ///                         LatticeFasterDecoder::GetLattice().
+  bool GetLattice(bool use_final_probs, bool redeterminize, CompactLattice *olat);
+  CompactLattice lat_;                            // the compact lattice we obtain
+  int32 last_get_lattice_frame_;                  // the last time we call GetLattice
+  unordered_map<Token *, int32> state_label_map_; // between Token and state_label
+  int32 state_label_avilable_idx_;     // we allocate a unique id for each Token
+  const TransitionModel &trans_model_; // keep it for determinization
+  std::vector<std::pair<StateId, size_t>> final_arc_list_; // keep final_arc
+  std::vector<std::pair<StateId, size_t>> final_arc_list_prev_;
+  // TODO use 2 vector to replace this map, since state_label is continuous
+  // in each frame, and we need 2 frames of them
+  unordered_map<int32, BaseFloat>
+      state_label_forward_prob_; // alpha for each state_label (Token)
+  // specific design for incremental GetLattice
+  bool GetRawLattice(Lattice *ofst, bool use_final_probs, int32 frame_begin,
+                     int32 frame_end, bool create_initial_state,
```

Comment:
I don't see how create_initial_state == false would make sense.  And you should document things like this as you write them.

Reply:
we do not create the initial state in the first chunk
we do not create the final state in the last chunk
comments added